src/src.rs

```diff
@@ -101,6 +103,108 @@ impl<'ctx> SourceGroup<'ctx> {
         )
     }
 
+    /// recursively get dependency names
+    fn get_deps(&self, packages: &HashSet<String>, excludes: &HashSet<String>) -> HashSet<String> {
+        let mut result = packages.clone();
+
+        match self.package {
+            Some(x) => {
+                if result.contains(x) {
+                    result.extend(HashSet::<String>::from_iter(self.dependencies.clone()));
+                    result = &result - &excludes;
+                }
+            }
+            None => {}
+        }
+
+        for file in &self.files {
+            match file {
+                SourceFile::Group(group) => {
+                    result.extend(group.get_deps(&result, excludes));
+                }
+                _ => {}
+            }
+        }
+
+        result
+    }
+
+    /// Get list of packages bassed on constraints
+    pub fn get_package_list(
+        &self,
+        sess: &Session,
+        packages: &HashSet<String>,
+        excludes: &HashSet<String>,
+        no_deps: bool,
+    ) -> HashSet<String> {
+        let mut result = HashSet::new();
+
+        if !packages.is_empty() {
+            result.extend(packages.clone());
+        } else {
+            result.insert(sess.manifest.package.name.to_string());
+        }
+
+        result = &result - &excludes;
+
+        if !no_deps {
+            let mut curr_length = 0;
+            while curr_length < result.len() {
+                curr_length = result.len();
+                result.extend(self.get_deps(&result, excludes));
+            }
+        }
+
+        result
+    }
+
+    /// Filter the sources, keeping only the ones that apply to the selected packages.
+    pub fn filter_packages(&self, packages: &HashSet<String>) -> Option<SourceGroup<'ctx>> {
+        let mut files = Vec::new();
+
+        match self.package {
+            Some(x) => {
+                if packages.contains(&x.to_string()) {
+                    files = self
+                        .files
+                        .iter()
+                        .filter_map(|file| match *file {
+                            SourceFile::Group(ref group) => group
+                                .filter_packages(packages)
+                                .map(|g| SourceFile::Group(Box::new(g))),
+                            ref other => Some(other.clone()),
+                        })
+                        .collect();
+                }
+            }
+            None => {
+                files = self
+                    .files
+                    .iter()
+                    .filter_map(|file| match *file {
+                        SourceFile::Group(ref group) => group
+                            .filter_packages(packages)
+                            .map(|g| SourceFile::Group(Box::new(g))),
+                        ref other => Some(other.clone()),
+                    })
+                    .collect();
+            }
+        }
```

Comment:
A `match` statement here leads to redundant code. I suggest to instead use a classic `if` with short-circuit evaluation:
```suggestion
        if self.package.is_none() || packages.contains(self.package.unwrap()) {
            files = self
                .files
                .iter()
                .filter_map(|file| match *file {
                    SourceFile::Group(ref group) => group
                        .filter_packages(packages)
                        .map(|g| SourceFile::Group(Box::new(g))),
                    ref other => Some(other.clone()),
                })
                .collect();
        }
```